dlt/common/schema/schema.py

```diff
@@ -177,6 +184,71 @@ def coerce_row(self, table_name: str, parent_table: str, row: StrAny) -> Tuple[D
+        schema_evolution_modes = resolve_single(self._settings.get("schema_evolution_settings", {}))
+        overide_modes = resolve_single(schema_evolution_settings_override)
+
+        # resolve to correct settings dict
+        settings = cast(TSchemaEvolutionModes, {**DEFAULT_SCHEMA_EVOLUTION_MODES, **schema_evolution_modes, **table_evolution_modes, **overide_modes})
+
+        return settings
+
+
+    def check_schema_update(self, parent_table: str, table_name: str, row: DictStrAny, partial_table: TPartialTableSchema, schema_evolution_settings_override: TSchemaEvolutionSettings) -> Tuple[DictStrAny, TPartialTableSchema]:
+        """Checks if schema update mode allows for the requested changes, filter row or reject update, depending on the mode"""
+
+        assert partial_table
+
+        # for now we defined the schema as new if there are no data columns defined
+        has_columns = self.has_data_columns
+        if not has_columns:
+            return row, partial_table
+
+        evolution_modes = self.resolve_evolution_settings_for_table(parent_table, table_name, schema_evolution_settings_override)
+
+        # default settings allow all evolutions, skipp all else
+        if evolution_modes == DEFAULT_SCHEMA_EVOLUTION_MODES:
+            return row, partial_table
+
+        table_exists = table_name in self.tables and len(self.tables[table_name].get("columns", {}))
+
+        # check case where we have a new table
+        if not table_exists:
+            if evolution_modes == "freeze-and-trim":
+                return None, None
+            if evolution_modes["table"] in ["freeze-and-discard", "freeze-and-trim"]:
+                return None, None
+            if evolution_modes["table"] == "freeze-and-raise":
+                raise SchemaFrozenException(f"Trying to add table {table_name} but new tables are frozen.")
+
+        # check columns
+        for item in list(row.keys()):
+            for item in list(row.keys()):
+                # if this is a new column for an existing table...
+                if table_exists and item not in self.tables[table_name]["columns"]:
+                    is_variant = item in partial_table["columns"] and partial_table["columns"][item].get("variant")
+                    if evolution_modes["column"] == "freeze-and-trim" or (is_variant and evolution_modes["column_variant"] == "freeze-and-trim"):
+                        row.pop(item)
+                        partial_table["columns"].pop(item)
+                    if evolution_modes["column"] == "freeze-and-discard" or (is_variant and evolution_modes["column_variant"] == "freeze-and-discard"):
+                        return None, None
+                    if evolution_modes["column"] == "freeze-and-raise" or (is_variant and evolution_modes["column_variant"] == "freeze-and-raise"):
+                        raise SchemaFrozenException(f"Trying to add column {item} to table {table_name}  but columns are frozen.")
```

Comment:
we really need a good exception for variant type that explains that variant column tried to be created but failed